Prevent snapshots to be mounted as system indices

tlrx opened the pull request:
System and hidden indices can be snapshotted and are therefore potential candidates to be mounted as searchable snapshot indices. As of today nothing prevents a snapshot to be mounted under an index name starting with `.` and this can lead to conflicting situations because searchable snapshot indices are read-only and Elasticsearch expects some system indices to be writable; because searchable snapshot indices will soon use an internal system index (#60522) to speed up recoveries and we should prevent the system index to be itself a searchable snapshot index (leading to some deadlock situation for recovery).

This pull request introduces a changes to prevent snapshots to be mounted under a system index.

tlrx committed:
Prevent snapshots to be mounted as system indices

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Snapshot/Restore)

fcofdez reviewed (approved):
LGTM

jaymode commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java:
```diff
@@ -132,6 +133,11 @@ protected void masterOperation(
         SearchableSnapshots.ensureValidLicense(licenseState);
 
+        final String mountedIndexName = request.mountedIndexName();
+        if (mountedIndexName.charAt(0) == '.') {
```
I think it would be best to pull in the `SystemIndices` class since we're in a transport action. We can bind the instance in `Node` and make it available for guice to inject. Then a simple modification to the the `SystemIndices` class could be made that adds a method to check the name (currently there is a method that takes an `Index` object but only looks at the name) against the defined set of system indices.

The primary reason for this is users could have data indices in 7.x that start with a `.` so we shouldn't break that for them.

gwbrown replied:
There are also hidden indices that start with a dot (e.g. `.watcher-history*`), which there's no reason to prevent being backed by a searchable snapshot. Using the `SystemIndices` class would avoid that issue as well.

jaymode replied:
I merged #61540 which adds the method in `SystemIndices` to check a string name without needing the `Index` object.

tlrx replied:
Thanks everybody. I'm waiting for #60522 to be merged and I'll update this PR.

dakrone commented:
As a drive-by comment, I'm slightly concerned that this will trip up users that are trying to mount data stream indices manually as searchable snapshots. Data stream backing indices start with a `.` and we don't currently have a way to add existing indices to a data stream, so I'm concerned a user that is manually closing an index and restoring it as a searchable version won't be able to do that.

Maybe it's not a big deal, but at the very least we should document this somewhere?

tlrx committed:
Merge branch 'master' into prevent-system-indices-to-be-mounted

tlrx committed:
Use systemIndices

tlrx commented:
@jaymode @gwbrown @dakrone Thanks for your valuable feedback. I reverted some changes so that we are now more permissive and only prevent registered system indices to be mounted as searchable snapshots. I think this addresses your concerns.

jaymode reviewed (approved):
LGTM. Thanks @tlrx

fcofdez commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java:
```diff
@@ -132,6 +136,11 @@ protected void masterOperation(
         SearchableSnapshots.ensureValidLicense(licenseState);
 
+        final String mountedIndexName = request.mountedIndexName();
+        if (systemIndices.isSystemIndex(mountedIndexName)) {
```
This looks better than the previous approach 👍

fcofdez commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsSystemIndicesIntegTests.java:
```diff
@@ -0,0 +1,124 @@
+                .setSettings(Settings.builder().put(IndexMetadata.SETTING_INDEX_HIDDEN, isHidden).build())
+        );
+
+        final int nbDocs = scaledRandomIntBetween(0, 100);
```
nit: we could use `createAndPopulateIndex` instead of manually creating and populating the index?

tlrx replied:
Yes we can - I pushed 3e173cb

fcofdez reviewed (approved):
LGTM

tlrx committed:
use createAndPopulateIndex()

tlrx committed:
Merge branch 'master' into prevent-system-indices-to-be-mounted

tlrx commented:
@elasticmachine run elasticsearch-ci/packaging-sample-windows

tlrx commented:
@elasticmachine run elasticsearch-ci/packaging-sample-windows

tlrx commented:
Thanks all!
